### PR TITLE
Reason for change: new hal api "wifi_startNeighborScan" implementation

### PR DESCRIPTION
**Pull request by VMLTTS**
Test Procedure:  use wifihal command options
                           eg : wifihal wifi_startNeighborScan 6 0 10 2 2412 2437

### CODE REVIEW

File: source/wifi/wifi_hal.c [hunk 1]
```diff
@@ -8425,6 +8452,31 @@ int main(int argc,char **argv)
...
         printf("wifi_devMacAddress=%s \t wifi_devAssociatedDeviceAuthentiationState=%d \t, wifi_devSignalStrength=%d \t,wifi_devTxRate=%d \t, wifi_devRxRate =%d \t\n ", mac_addr,output_struct.wifi_devAssociatedDeviceAuthentiationState,output_struct.wifi_devSignalStrength,output_struct.wifi_devTxRate,output_struct.wifi_devRxRate);
     }
+	
+   if(strstr(argv[1],"wifi_startNeighborScan")!=NULL)
+    {
+	int args_index = 0;
+        if(argc <= 5 )
+        {
+
+    	 printf("Insufficient arguments , usage :: \n");
```
Comment (manigandanj) [inline source/wifi/wifi_hal.c]: Make sure indent are uniform.

File: source/wifi/wifi_hal.c [hunk 2]
```diff
@@ -7200,8 +7200,35 @@ INT wifi_getApAssociatedDeviceTidStatsResult(INT radioIndex,  mac_address_t *cli
...
+    }
+
+     /* If channels are not specified use default driver params */
+    if (chan_num)
+    {
+      uint32_t    chan_index;
+
+       for (chan_index = 0; chan_index < chan_num; chan_index++)
+           {   
+     		snprintf(cmd,sizeof(cmd),  "iw dev %s%d scan -u freq %u passive >> /tmp/%u.txt ", RADIO_PREFIX, apIndex%2, chan_list[chan_index],chan_list[chan_index]);
```
Comment (mbajorski) [inline source/wifi/wifi_hal.c]: Please use wifi_getApRadioIndex instead of apIndex%2 to make more generic implementation.

If we use this wifihal on different platform with different number of radios this function might not work as expected

File: source/wifi/wifi_hal.c [hunk 3]
```diff
@@ -104,6 +104,7 @@
 #define BW_FNAME "/nvram/bw_file.txt"
 
 #define PS_MAX_TID 16
+BOOL G_BS_ENABLE = FALSE ;
```
Comment (mbajorski) [inline source/wifi/wifi_hal.c]: Please create new PR for this change.
This is not related to wifi_startNeighborScan

**Review by mbajorski — CHANGES_REQUESTED** [review]

File: source/wifi/wifi_hal.c (hunk 1 above)
Comment (VMLTTS) [reply source/wifi/wifi_hal.c]: updated

File: source/wifi/wifi_hal.c (hunk 2 above)
Comment (VMLTTS) [reply source/wifi/wifi_hal.c]: used wifi_getApRadioIndex for calculating the radio index

File: source/wifi/wifi_hal.c (hunk 3 above)
Comment (VMLTTS) [reply source/wifi/wifi_hal.c]: removed the BS related changes in the current PR

File: source/wifi/wifi_hal.c [hunk 4]
```diff
@@ -7200,11 +7199,32 @@ INT wifi_getApAssociatedDeviceTidStatsResult(INT radioIndex,  mac_address_t *cli
...
+    }
+     /* If channels are not specified use default driver params */
+    if (chan_num)
+    {
+        uint32_t    chan_index;
+        for (chan_index = 0; chan_index < chan_num; chan_index++)
+        {   
+     	    snprintf(cmd,sizeof(cmd),  "iw dev %s%d scan -u freq %u passive >> /tmp/%u.txt ", RADIO_PREFIX, Radio_Index, chan_list[chan_index],chan_list[chan_index]);
+	    system(cmd);
+        }
```
Comment (mbajorski) [inline source/wifi/wifi_hal.c:7229]: Please fix indentations

File: source/wifi/wifi_hal.c [hunk 5]
```diff
@@ -7200,11 +7199,32 @@ INT wifi_getApAssociatedDeviceTidStatsResult(INT radioIndex,  mac_address_t *cli
...
+     }else {
```
Comment (mbajorski) [inline source/wifi/wifi_hal.c]: add space before else

File: source/wifi/wifi_hal.c [hunk 6]
```diff
@@ -7200,11 +7199,32 @@ INT wifi_getApAssociatedDeviceTidStatsResult(INT radioIndex,  mac_address_t *cli
...
+    {   
+        return RETURN_OK;
+    }
+     /* If channels are not specified use default driver params */
+    if (chan_num)
+    {
+        uint32_t    chan_index;
+        for (chan_index = 0; chan_index < chan_num; chan_index++)
+        {   
+     	    snprintf(cmd,sizeof(cmd),  "iw dev %s%d scan -u freq %u passive >> /tmp/%u.txt ", RADIO_PREFIX, Radio_Index, chan_list[chan_index],chan_list[chan_index]);
```
Comment (mbajorski) [inline source/wifi/wifi_hal.c]: -u freq accepts frequency
chan_list is list of channels not list of frequencies

File: source/wifi/wifi_hal.c [hunk 7]
```diff
@@ -8425,6 +8445,27 @@ int main(int argc,char **argv)
...
+    	    printf("Insufficient arguments , usage :: \n");  
+            printf(" wifihal wifi_startNeighborScan [apindex (0-15)] [scanmode (0 to 4)] [dwelltime] [channel_listcount] [ channel-list]... \n");    
+	    exit(-1);
+        }
+        INT apindex = atoi(argv[2]);
+        wifi_neighborScanMode_t scan_mode = atoi(argv[3]);
+        INT dwell_time = atoi(argv[4]);
+        UINT chan_num = atoi(argv[5]);
+	UINT *chan_list = (UINT *) malloc((argc-5)*sizeof(UINT));
+	for( args_index=0 ; args_index < argc-6 ; args_index++ ){
```
Comment (mbajorski) [inline source/wifi/wifi_hal.c]: Please align spaces to previous "for" loop implementation

File: source/wifi/wifi_hal.c [hunk 8]
```diff
@@ -7200,11 +7199,32 @@ INT wifi_getApAssociatedDeviceTidStatsResult(INT radioIndex,  mac_address_t *cli
 
 INT wifi_startNeighborScan(INT apIndex, wifi_neighborScanMode_t scan_mode, INT dwell_time, UINT chan_num, UINT *chan_list)
 {
-    // TODO Implement me!
+    INT status = RETURN_ERR;
+    UINT index;
+    INT Radio_Index;
+    char cmd[256]={0};
+    WIFI_ENTRY_EXIT_DEBUG("Inside %s:%d\n",__func__, __LINE__);
+    wifi_getApRadioIndex(apIndex,&Radio_Index);
```
Comment (mbajorski) [inline source/wifi/wifi_hal.c]: please check return code

File: source/wifi/wifi_hal.c (hunk 6 above)
Comment (mbajorski) [reply source/wifi/wifi_hal.c]: What happens if we call this function multiple times?

Comment (mbajorski) [reply source/wifi/wifi_hal.c]: if you save something in /tmp/ please give meaningful filename prefix to know from where we have this file

File: source/wifi/wifi_hal.c [hunk 9]
```diff
@@ -8425,6 +8445,27 @@ int main(int argc,char **argv)
...
+        {
+    	    printf("Insufficient arguments , usage :: \n");  
+            printf(" wifihal wifi_startNeighborScan [apindex (0-15)] [scanmode (0 to 4)] [dwelltime] [channel_listcount] [ channel-list]... \n");    
+	    exit(-1);
+        }
+        INT apindex = atoi(argv[2]);
+        wifi_neighborScanMode_t scan_mode = atoi(argv[3]);
+        INT dwell_time = atoi(argv[4]);
+        UINT chan_num = atoi(argv[5]);
+	UINT *chan_list = (UINT *) malloc((argc-5)*sizeof(UINT));
```
Comment (mbajorski) [inline source/wifi/wifi_hal.c]: please align indentations

File: source/wifi/wifi_hal.c [hunk 10]
```diff
@@ -8425,6 +8445,27 @@ int main(int argc,char **argv)
         mac_addr_ntoa(mac_addr,output_struct.wifi_devMacAddress);
         printf("wifi_devMacAddress=%s \t wifi_devAssociatedDeviceAuthentiationState=%d \t, wifi_devSignalStrength=%d \t,wifi_devTxRate=%d \t, wifi_devRxRate =%d \t\n ", mac_addr,output_struct.wifi_devAssociatedDeviceAuthentiationState,output_struct.wifi_devSignalStrength,output_struct.wifi_devTxRate,output_struct.wifi_devRxRate);
     }
+	
+   if(strstr(argv[1],"wifi_startNeighborScan")!=NULL)
+    {
+	int args_index = 0;
+        if(argc <= 5 )
```
Comment (mbajorski) [inline source/wifi/wifi_hal.c]: Fix spaces inside of "if"

File: source/wifi/wifi_hal.c [hunk 11]
```diff
@@ -7200,11 +7199,32 @@ INT wifi_getApAssociatedDeviceTidStatsResult(INT radioIndex,  mac_address_t *cli
 
 INT wifi_startNeighborScan(INT apIndex, wifi_neighborScanMode_t scan_mode, INT dwell_time, UINT chan_num, UINT *chan_list)
 {
-    // TODO Implement me!
+    INT status = RETURN_ERR;
+    UINT index;
```
Comment (mbajorski) [inline source/wifi/wifi_hal.c]: Unused variable

File: source/wifi/wifi_hal.c [hunk 12]
```diff
@@ -7200,11 +7199,32 @@ INT wifi_getApAssociatedDeviceTidStatsResult(INT radioIndex,  mac_address_t *cli
 
 INT wifi_startNeighborScan(INT apIndex, wifi_neighborScanMode_t scan_mode, INT dwell_time, UINT chan_num, UINT *chan_list)
 {
-    // TODO Implement me!
+    INT status = RETURN_ERR;
```
Comment (mbajorski) [inline source/wifi/wifi_hal.c]: Unused variable

File: source/wifi/wifi_hal.c [hunk 13]
```diff
@@ -7200,11 +7199,32 @@ INT wifi_getApAssociatedDeviceTidStatsResult(INT radioIndex,  mac_address_t *cli
...
+        for (chan_index = 0; chan_index < chan_num; chan_index++)
+        {   
+     	    snprintf(cmd,sizeof(cmd),  "iw dev %s%d scan -u freq %u passive >> /tmp/%u.txt ", RADIO_PREFIX, Radio_Index, chan_list[chan_index],chan_list[chan_index]);
+	    system(cmd);
+        }
+     }else { 
+         snprintf(cmd, sizeof(cmd),  "iw dev %s%d scan passive >> /tmp/scan_results.txt ", RADIO_PREFIX, Radio_Index );
+         system(cmd);
+     }
     return RETURN_OK;
```
Comment (mbajorski) [inline source/wifi/wifi_hal.c]: You use WIFI_ENTRY_EXIT_DEBUG on entry but there is missing WIFI_ENTRY_EXIT_DEBUG on exit

File: source/wifi/wifi_hal.c [hunk 14]
```diff
@@ -7200,11 +7199,32 @@ INT wifi_getApAssociatedDeviceTidStatsResult(INT radioIndex,  mac_address_t *cli
 
 INT wifi_startNeighborScan(INT apIndex, wifi_neighborScanMode_t scan_mode, INT dwell_time, UINT chan_num, UINT *chan_list)
 {
-    // TODO Implement me!
+    INT status = RETURN_ERR;
+    UINT index;
+    INT Radio_Index;
+    char cmd[256]={0};
```
Comment (mbajorski) [inline source/wifi/wifi_hal.c]: You can use MAX_CMD_SIZE instead of fixed number 256

File: source/wifi/wifi_hal.c [hunk 15]
```diff
@@ -8425,6 +8445,27 @@ int main(int argc,char **argv)
...
+        }
+        INT apindex = atoi(argv[2]);
+        wifi_neighborScanMode_t scan_mode = atoi(argv[3]);
+        INT dwell_time = atoi(argv[4]);
+        UINT chan_num = atoi(argv[5]);
+	UINT *chan_list = (UINT *) malloc((argc-5)*sizeof(UINT));
+	for( args_index=0 ; args_index < argc-6 ; args_index++ ){
+      	    chan_list[args_index] = (UINT)atoi(argv[args_index+6]);
+        }
+        wifi_startNeighborScan(apindex, scan_mode, dwell_time, chan_num, chan_list);
```
Comment (mbajorski) [inline source/wifi/wifi_hal.c]: please print return code from this wifi hal function

**Review by mbajorski — CHANGES_REQUESTED** [review]

File: source/wifi/wifi_hal.c (hunk 4 above)
Comment (VMLTTS) [reply source/wifi/wifi_hal.c:7229]: done

File: source/wifi/wifi_hal.c (hunk 10 above)
Comment (VMLTTS) [reply source/wifi/wifi_hal.c]: done

File: source/wifi/wifi_hal.c (hunk 8 above)
Comment (VMLTTS) [reply source/wifi/wifi_hal.c]: Done

File: source/wifi/wifi_hal.c (hunk 6 above)
Comment (VMLTTS) [reply source/wifi/wifi_hal.c]: updated the logic to input the channel number and assigned proper naming to the output files

File: source/wifi/wifi_hal.c (hunk 7 above)
Comment (VMLTTS) [reply source/wifi/wifi_hal.c]: done

File: source/wifi/wifi_hal.c (hunk 11 above)
Comment (VMLTTS) [reply source/wifi/wifi_hal.c]: fixed

File: source/wifi/wifi_hal.c (hunk 12 above)
Comment (VMLTTS) [reply source/wifi/wifi_hal.c]: fixed

File: source/wifi/wifi_hal.c (hunk 13 above)
Comment (VMLTTS) [reply source/wifi/wifi_hal.c]: done

File: source/wifi/wifi_hal.c (hunk 14 above)
Comment (VMLTTS) [reply source/wifi/wifi_hal.c]: done

File: source/wifi/wifi_hal.c (hunk 15 above)
Comment (VMLTTS) [reply source/wifi/wifi_hal.c]: done

File: source/wifi/wifi_hal.c [hunk 16]
```diff
@@ -7199,30 +7199,40 @@ INT wifi_getApAssociatedDeviceTidStatsResult(INT radioIndex,  mac_address_t *cli
...
+                printf("Failed to get frequency from the chan_to_freq");
+                return RETURN_ERR;
+            }
+            snprintf(cmd,sizeof(cmd),  "iw dev %s%d scan -u freq %d passive > /tmp/NeighborScan_CHN%d_results.txt ", RADIO_PREFIX, Radio_Index,freq,chan_list[chan_index]);
+            system(cmd);
         }
-     }else { 
-         snprintf(cmd, sizeof(cmd),  "iw dev %s%d scan passive >> /tmp/scan_results.txt ", RADIO_PREFIX, Radio_Index );
+     } else {
+         snprintf(cmd, sizeof(cmd),  "iw dev %s%d scan passive > /tmp/NeighborScan_results.txt ", RADIO_PREFIX, Radio_Index );
```
Comment (mbajorski) [inline source/wifi/wifi_hal.c]: Please mention radio index inf file name to not collide with other radio indexes
Remove space on the end

File: source/wifi/wifi_hal.c (hunk 10 above)
Comment (mbajorski) [reply source/wifi/wifi_hal.c]: it is not done

File: source/wifi/wifi_hal.c (hunk 4 above)
Comment (mbajorski) [reply source/wifi/wifi_hal.c:7229]: Add radio index in temp file

File: source/wifi/wifi_hal.c [hunk 17]
```diff
@@ -7199,30 +7199,40 @@ INT wifi_getApAssociatedDeviceTidStatsResult(INT radioIndex,  mac_address_t *cli
...
         uint32_t    chan_index;
         for (chan_index = 0; chan_index < chan_num; chan_index++)
-        {   
-     	    snprintf(cmd,sizeof(cmd),  "iw dev %s%d scan -u freq %u passive >> /tmp/%u.txt ", RADIO_PREFIX, Radio_Index, chan_list[chan_index],chan_list[chan_index]);
-	    system(cmd);
+        {
+            int freq = 0;
+            if( chan_to_freq(Radio_Index,chan_list[chan_index],&freq) == RETURN_ERR )
+            {
+                printf("Failed to get frequency from the chan_to_freq");
```
Comment (mbajorski) [inline source/wifi/wifi_hal.c]: Please use wifi_dbg_printf instead of printf and add \n on the end

Comment (VMLTTS) [reply source/wifi/wifi_hal.c]: updated

File: source/wifi/wifi_hal.c (hunk 16 above)
Comment (VMLTTS) [reply source/wifi/wifi_hal.c]: updated the file name to NeighborScan_RADIO%d_results.txt

**Comment by scthunderbolt** [comment]
Closed due to inactivity, please restore if still needed.
